polymorphize: unevaluated constants

davidtwco opened the pull request:
This PR makes polymorphization visit the promoted MIR of unevaluated constants with available promoted MIR instead of visiting the substitutions of that constant - which will mark all of the generic parameters as used; in addition polymorphization will now visit non-promoted unevaluated constants rather than visit their substs.

r? @lcnr

davidtwco committed:
polymorphize: visit promoted MIR

This commit makes polymorphization visited the MIR of unevaluated
constants with available promoted MIR instead of visiting the
substitutions of that constant - which will mark all of the generic
parameters as used.

Signed-off-by: David Wood <david@davidtw.co>

lcnr commented on src/librustc_mir/monomorphize/polymorphize.rs:
```diff
@@ -245,6 +245,13 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
                 self.unused_parameters.clear(param.index);
                 false
             }
+            ty::ConstKind::Unevaluated(_, _, Some(p)) => {
```
We actually have the same problem for other unevaluated constants:
```rust
#![feature(lazy_normalization_consts)]

fn test<T>() {
    let x = [0; 3 + 4];
}

pub fn caller() {
    test::<String>();
    test::<Vec<String>>();
}
```
But afaict that doesn't concern stable rn, not completely sure if we can end up with cycle errors here, but I think we should be fine and able to use `unused_generic_params` here (or we just have to restrict this to `DefKind::AnonConst` for now)

davidtwco replied:
Fixed!

lcnr reviewed (approved):
In case you don't want to look into the more general case here,

r=me

davidtwco committed:
polymorphize: non-promoted unevaluated constants

This commit makes polymorphization visit non-promoted unevaluated
constants rather than visit their substs directly.

Signed-off-by: David Wood <david@davidtw.co>

lcnr reviewed (commented):
nice :+1: once the PR title and description has been updated,

r=me

davidtwco commented:
@bors r=lcnr

bors commented:
:pushpin: Commit d97f89b1a6add41b6ac3c0d69cbe82d15483a0d6 has been approved by `lcnr`

<!-- @bors r=lcnr d97f89b1a6add41b6ac3c0d69cbe82d15483a0d6 -->
<!-- homu: {"type":"Approved","sha":"d97f89b1a6add41b6ac3c0d69cbe82d15483a0d6","approver":"lcnr"} -->

bors commented:
:hourglass: Testing commit d97f89b1a6add41b6ac3c0d69cbe82d15483a0d6 with merge 3f091baba4fa656adb4c1a57b64aa831002d801d...
<!-- homu: {"type":"BuildStarted","head_sha":"d97f89b1a6add41b6ac3c0d69cbe82d15483a0d6","merge_sha":"3f091baba4fa656adb4c1a57b64aa831002d801d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/961996180), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35869)
Approved by: lcnr
Pushing 3f091baba4fa656adb4c1a57b64aa831002d801d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35869","checks-actions":"https://github.com/rust-lang-ci/rust/runs/961996180"},"merge_sha":"3f091baba4fa656adb4c1a57b64aa831002d801d"} -->
